### PR TITLE
Enable FTL on Windows

### PR DESCRIPTION
**Pull request by iangrunert**
<pre>
Enable FTL on Windows
<a href="https://bugs.webkit.org/show_bug.cgi?id=145366">https://bugs.webkit.org/show_bug.cgi?id=145366</a>

Reviewed by NOBODY (OOPS!).

Enabling webassembly at build time and disabling at runtime, as
FTL makes assumptions that webassembly is enabled. Should be able
to apply a cleaner patch once <a href="https://bugs.webkit.org/show_bug.cgi?id=222315">https://bugs.webkit.org/show_bug.cgi?id=222315</a>
lands (awaiting PR review).

The BBQ JIT won't work on Windows even after the changes for
enabling webassembly on Windows land - it requires changes to GPR
assignment so it can use callee saves. With two fewer GPRs to
work with it runs out of GPRs to assign when running benchmarks.

Currently segfaulting during JetStream2 when running stanford
crypto-aes (on most but not all runs, it occasionally works).

FTL is disabled on Windows for EnumeratorNextUpdateIndexAndMode.
It crashes on Windows with a segfault on a simple test with
JSC_useConcurrentJIT=false. If converted from UGPRPair return
type back to JSValue it works - so either something in the FTL
lowering, or the register allocator is having problems due to the
extra return register required.

```
function getMappedArguments() { return arguments; }

function forIn(object) {
    var keys = [];
    for (var key in object)
        keys.push(key);
    return keys;
}

noInline(forIn);

(function() {
    var mappedArguments = getMappedArguments(0, 1, 2);
    for (var i = 0; i < 1e4; ++i) {
        forIn(mappedArguments);
    }
    print('done');
})();
```

* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/air/AirCCallSpecial.cpp: (JSC::B3::Air::CCallSpecial::generate):
* Source/JavaScriptCore/b3/air/AirCCallingConvention.cpp:
* Source/JavaScriptCore/b3/air/AirCode.h: (JSC::B3::Air::Code::regsInPriorityOrder const):
(JSC::B3::Air::Code::regsInPriorityOrderImpl):
* Source/JavaScriptCore/b3/air/AirPrintSpecial.h:
* Source/JavaScriptCore/b3/air/opcode_generator.rb:
* Source/JavaScriptCore/b3/air/testair.cpp:
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/b3/testb3_3.cpp: (correctSqrt):
(testStorePartial8BitRegisterOnX86):
* Source/JavaScriptCore/b3/testb3_4.cpp: (testLoadFromFramePointer):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp: (JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp: (JSC::FTL::DFG::LowerDFGToB3::compileDelBy):
(JSC::FTL::DFG::LowerDFGToB3::compileMultiGetByOffset): (JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLOSRExitCompiler.cpp: (JSC::FTL::compileStub):
* Source/JavaScriptCore/ftl/FTLOperations.h:
* Source/JavaScriptCore/ftl/FTLOutput.cpp: (JSC::FTL::Output::baseIndex):
* Source/JavaScriptCore/ftl/FTLSlowPathCall.cpp: (JSC::FTL::SlowPathCallContext::SlowPathCallContext):
* Source/JavaScriptCore/jit/RegisterSet.cpp: (JSC::RegisterSetBuilder::ftlCalleeSaveRegisters):
* Source/JavaScriptCore/llint/LLIntData.cpp: (JSC::LLInt::initialize):
* Source/JavaScriptCore/llint/LLIntThunks.cpp: (JSC::LLInt::untaggedPtr):
(JSC::LLInt::inPlaceInterpreterEntryThunk):
(JSC::LLInt::inPlaceInterpreterEntryThunkSIMD):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/offlineasm/x86.rb:
* Source/JavaScriptCore/runtime/Options.cpp: (JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp: (JSC::Wasm::B3IRGenerator::addI64TruncUF64):
(JSC::Wasm::B3IRGenerator::addI64TruncUF32):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp: (JSC::Wasm::BBQJIT::RegisterBinding::none):
(JSC::Wasm::BBQJIT::addI32Rotl):
(JSC::Wasm::BBQJIT::addI64Rotl):
(JSC::Wasm::BBQJIT::addF32ConvertUI32):
(JSC::Wasm::BBQJIT::addF32ConvertUI64):
(JSC::Wasm::BBQJIT::addF64ConvertUI32):
(JSC::Wasm::BBQJIT::addF64ConvertUI64):
(JSC::Wasm::BBQJIT::addF32Copysign):
(JSC::Wasm::BBQJIT::addF64Copysign):
(JSC::Wasm::BBQJIT::addF32Abs):
(JSC::Wasm::BBQJIT::addF64Abs):
(JSC::Wasm::BBQJIT::addF32Neg):
(JSC::Wasm::BBQJIT::addF64Neg):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp: (JSC::LLInt::WASM_IPINT_EXTERN_CPP_DECL):
* Source/cmake/OptionsJSCOnly.cmake:
* Source/cmake/OptionsWin.cmake:
</pre>

https://github.com/WebKit/WebKit/commit/7ea2d9802c5b581c103e936813dc39761634202d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23220 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1309 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24340 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/25562 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21456 "Built successfully") 
| | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2827 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23758 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/25562 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23460 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/2827 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/14/builds/24340 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25983 "Hash 7ea2d980 for PR 19073 does not build (failure)") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/2827 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/14/builds/24340 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/25983 "Hash 7ea2d980 for PR 19073 does not build (failure)") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/20209 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/2827 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/14/builds/24340 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/25983 "Hash 7ea2d980 for PR 19073 does not build (failure)") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22583 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/713 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/16/builds/23758 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/29960 "Hash 7ea2d980 for PR 19073 does not build (failure)") | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/651 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/14/builds/24340 "Hash 7ea2d980 for PR 19073 does not build (failure)") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/29960 "Hash 7ea2d980 for PR 19073 does not build (failure)") | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1123 "Hash 7ea2d980 for PR 19073 does not build (failure)") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/29912 "Built successfully") | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/926 "Hash 7ea2d980 for PR 19073 does not build (failure)") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6086 "Passed tests") |

### CODE REVIEW

File: Source/JavaScriptCore/b3/air/AirCCallSpecial.cpp [hunk 1]
```diff
@@ -130,6 +130,10 @@ void CCallSpecial::reportUsedRegisters(Inst&, const RegisterSetBuilder&)
 
 CCallHelpers::Jump CCallSpecial::generate(Inst& inst, CCallHelpers& jit, GenerationContext&)
 {
+#if OS(WINDOWS) && CPU(X86_64)
+    jit.sub64(CCallHelpers::TrustedImm32(4 * sizeof(int64_t)), X86Registers::esp);
+#endif
```
Comment (Constellation) [inline Source/JavaScriptCore/b3/air/AirCCallSpecial.cpp:135]: We should not have difference between Windows and others.

File: Source/JavaScriptCore/b3/air/AirCCallSpecial.cpp [hunk 2]
```diff
@@ -147,6 +151,11 @@ CCallHelpers::Jump CCallSpecial::generate(Inst& inst, CCallHelpers& jit, Generat
         RELEASE_ASSERT_NOT_REACHED();
         break;
     }
+
+#if OS(WINDOWS) && CPU(X86_64)
+    jit.add64(CCallHelpers::TrustedImm32(4 * sizeof(int64_t)), X86Registers::esp);
+#endif
```
Comment (Constellation) [inline Source/JavaScriptCore/b3/air/AirCCallSpecial.cpp:157]: Ditto.

File: Source/JavaScriptCore/b3/air/AirCCallingConvention.cpp [hunk 3]
```diff
@@ -83,7 +83,14 @@ void marshallCCallArgument(Vector<Arg> &result, unsigned& gpArgumentCount, unsig
...
         return;
     case FP:
+#if OS(WINDOWS) && CPU(X86_64)
+        // On Windows, arguments map to designated registers based on the argument positions,
+        // even when there are interlaced scalar and floating point arguments.
+        UNUSED_PARAM(fpArgumentCount);
+        marshallCCallArgumentImpl<FPRInfo>(result, gpArgumentCount, stackOffset, child);
+#else
         marshallCCallArgumentImpl<FPRInfo>(result, fpArgumentCount, stackOffset, child);
+#endif
```
Comment (Constellation) [inline Source/JavaScriptCore/b3/air/AirCCallingConvention.cpp:93]: We should not have difference between Windows and others.

File: Source/JavaScriptCore/b3/air/testair.cpp [hunk 4]
```diff
@@ -232,6 +232,7 @@ void testShuffleSimpleShift()
     CHECK(things[4] == 4);
 }
 
+#if !OS(WINDOWS)
```
Comment (Constellation) [inline Source/JavaScriptCore/b3/air/testair.cpp:235]: Why?

File: Source/JavaScriptCore/b3/air/testair.cpp [hunk 5]
```diff
@@ -449,6 +451,7 @@ void testShuffleBroadcastAllRegs()
         CHECK(thing == 35);
 }
 
+#if !OS(WINDOWS)
```
Comment (Constellation) [inline Source/JavaScriptCore/b3/air/testair.cpp:454]: We should not have difference between Windows and others.

File: Source/JavaScriptCore/b3/air/testair.cpp [hunk 6]
```diff
@@ -2643,14 +2647,18 @@ void run(const char* filter)
 
     RUN(testShuffleSimpleSwap());
     RUN(testShuffleSimpleShift());
+#if !OS(WINDOWS)
     RUN(testShuffleLongShift());
     RUN(testShuffleLongShiftBackwards());
+#endif
```
Comment (Constellation) [inline Source/JavaScriptCore/b3/air/testair.cpp:2653]: We should not have difference between Windows and others.

File: Source/JavaScriptCore/b3/air/testair.cpp [hunk 7]
```diff
@@ -2643,14 +2647,18 @@ void run(const char* filter)
...
     RUN(testShuffleSimpleRotate());
     RUN(testShuffleSimpleBroadcast());
     RUN(testShuffleBroadcastAllRegs());
+#if !OS(WINDOWS)
     RUN(testShuffleTreeShift());
     RUN(testShuffleTreeShiftBackward());
     RUN(testShuffleTreeShiftOtherBackward());
+#endif
```
Comment (Constellation) [inline Source/JavaScriptCore/b3/air/testair.cpp:2661]: We should not have difference between Windows and others.

File: Source/JavaScriptCore/b3/testb3_3.cpp [hunk 8]
```diff
@@ -2287,7 +2287,7 @@ void testFloorArgWithEffectfulDoubleConversion(float a)
 
 double correctSqrt(double value)
 {
-#if CPU(X86) || CPU(X86_64)
+#if (CPU(X86) || CPU(X86_64)) && !COMPILER(MSVC)
```
Comment (Constellation) [inline Source/JavaScriptCore/b3/testb3_3.cpp:2290]: We should not have difference between Windows and others.

File: Source/JavaScriptCore/b3/testb3_3.cpp [hunk 9]
```diff
@@ -3501,6 +3501,8 @@ void testStore8Imm()
 
 void testStorePartial8BitRegisterOnX86()
 {
+    // No regT6 on Windows
+#if !OS(WINDOWS)
```
Comment (Constellation) [inline Source/JavaScriptCore/b3/testb3_3.cpp:3505]: We should not have difference between Windows and others.

File: Source/JavaScriptCore/ftl/FTLCapabilities.cpp [hunk 10]
```diff
@@ -464,6 +463,14 @@ inline CapabilityLevel canCompile(Node* node)
...
         break;
 
+    case EnumeratorNextUpdateIndexAndMode:
+        // This doesn't work on Windows, but works elsewhere
+#if OS(WINDOWS)
+        return CannotCompile;
+#else
+        break;
+#endif
+
```
Comment (Constellation) [inline Source/JavaScriptCore/ftl/FTLCapabilities.cpp:473]: We should not have difference between Windows and others.

File: Source/JavaScriptCore/ftl/FTLSlowPathCall.cpp [hunk 11]
```diff
@@ -58,6 +59,10 @@ SlowPathCallContext::SlowPathCallContext(
         
     m_offsetToSavingArea =
         (std::max(m_numArgs, NUMBER_OF_ARGUMENT_REGISTERS) - NUMBER_OF_ARGUMENT_REGISTERS) * wordSize;
+
+#if OS(WINDOWS)
+    m_offsetToSavingArea = std::max(m_offsetToSavingArea, maxFrameExtentForSlowPathCall);
+#endif
```
Comment (Constellation) [inline Source/JavaScriptCore/ftl/FTLSlowPathCall.cpp:65]: We should not have difference between Windows and others.

File: Source/JavaScriptCore/jit/RegisterSet.cpp [hunk 12]
```diff
@@ -320,6 +320,17 @@ RegisterSet RegisterSetBuilder::ftlCalleeSaveRegisters()
...
+    result.add(GPRInfo::regCS1, IgnoreVectors);
+    result.add(GPRInfo::regCS2, IgnoreVectors);
+    result.add(GPRInfo::regCS3, IgnoreVectors);
+    static_assert(GPRInfo::regCS4 == GPRInfo::jitDataRegister);
+    static_assert(GPRInfo::regCS5 == GPRInfo::numberTagRegister);
+    static_assert(GPRInfo::regCS6 == GPRInfo::notCellMaskRegister);
+    result.add(GPRInfo::regCS4, IgnoreVectors);
+    result.add(GPRInfo::regCS5, IgnoreVectors);
+    result.add(GPRInfo::regCS6, IgnoreVectors);
 #elif CPU(ARM64)
```
Comment (Constellation) [inline Source/JavaScriptCore/jit/RegisterSet.cpp:334]: We should not have difference between Windows and others.

**Comment by iangrunert** [comment]
@Constellation This PR predates SysV ABI changes (October last year), so there's a lot of dead code in here.

**Pull request closed without merging**